### PR TITLE
Add hot reload

### PR DESCRIPTION
**Pull request by bb-face**
PR for the issue: https://github.com/NEARBuilders/near-bos-webcomponent/issues/20

Videos of the test:

https://github.com/NEARBuilders/near-bos-webcomponent/assets/152608874/343600d9-a347-4166-9011-998e8497f688

https://github.com/NEARBuilders/near-bos-webcomponent/assets/152608874/e9e8d690-16f6-4404-817f-e5f85959965d

https://github.com/NEARBuilders/near-bos-webcomponent/assets/152608874/85dba1cf-08df-436c-ae0e-d09f9a455fa6

https://github.com/NEARBuilders/near-bos-webcomponent/assets/152608874/0a8f37aa-d405-42c7-b1ed-411e4bf78501

### CODE REVIEW

**Review by petersalomonsen — CHANGES_REQUESTED** [review]
Quite a bit of changed formatting of the code, without any real changes. And also added lines with comments. A bit confusing when reviewing and trying to find the actual change.

Would it be possible to make the diff a bit clearer on what has actually changed?

File: src/contexts/hotReloadContext.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+import { createContext } from "react";
+
+export const HotReloadContext = createContext(false);
```
[the remark this answers is not in the document]
Comment (petersalomonsen) [reply src/contexts/hotReloadContext.js:3]: yes, shouldn't hot reload be configurable in some other way than directly in the source here? With this approach, how can hot-reload be configured when using the near-bos-webcomponents npm package?

**Comment by petersalomonsen** [comment]
Also see my comment here about the redirect map

https://github.com/NEARBuilders/near-bos-webcomponent/issues/20#issuecomment-2183714650

File: src/contexts/hotReloadContext.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (bb-face) [reply src/contexts/hotReloadContext.js]: @elliotBraem I tackle your comment in the last commit: https://github.com/NEARBuilders/near-bos-webcomponent/commit/4eff90140abed85a4a52b186aa0022ddc86a9e24

**Review by petersalomonsen — CHANGES_REQUESTED** [review]
Would it be possible to add a test for triggering the hot reload?

File: src/utils/redirectMap.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  const [hotReload, setHotReload] = useState(enableHotReload);
+  const [devJson, setDevJson] = useState({
+    components: {},
+    data: {},
+  });
+
+  useEffect(() => {
+    (async () => {
+      if (hotReload) {
+        const socket = io(process.env.BOS_LOADER_WS || "ws://127.0.0.1:8080", {
```
Comment (petersalomonsen) [inline src/utils/redirectMap.js:18]: Is `process.env` available here in the browser context?

Also, must it be only 127.0.0.1? If developing in github codespaces, you get dedicated host names with https.

Comment (bb-face) [reply src/utils/redirectMap.js]: Fixed here: https://github.com/NEARBuilders/near-bos-webcomponent/commit/d1a2839f63473b59cd49b85fe7ea9384eb439f3b

Comment (bb-face) [reply src/utils/redirectMap.js]: Fixed in this commit: https://github.com/NEARBuilders/near-bos-webcomponent/pull/26/commits/e1af6c47d1a6c67e9afaea9c14c0d4ec3665e26b

**Comment by elliotBraem** [comment]
bos-workspace is ready to support this:

https://github.com/NEARBuilders/bos-workspace/pull/118

**Review by petersalomonsen — APPROVED** [review]

**Review by petersalomonsen — CHANGES_REQUESTED** [review]
We also need the test for hot reload to ensure that what is implemented here will not be broken by someone else later. Also we need the test as a demonstration for how the hot reload works, and as a guide on how to use it.

File: src/utils/redirectMap.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      {children}
+    </HotReloadContext.Provider>
+  );
+};
+
+export function useRedirectMap() {
+  const enableHotReload = useContext(HotReloadContext);
+
+  const [hotReload, setHotReload] = useState(enableHotReload);
+  const [devJson, setDevJson] = useState({
```
Comment (elliotBraem) [inline src/utils/redirectMap.js:20]: useState({})

**Comment by elliotBraem** [comment]
@petersalomonsen yeah, good point on a test for validating hot reload. 

Since the socket server for hot reload needs to be on the same port as the app, it's been difficult to write a test for this -- we've tried mocking web socket or proxying requests, but it was challenging -- I'm not sure, @bb-face may have gotten closer to figuring it out?

Regardless, I'm wondering if this test belongs in bos-workspace repository instead? Or maybe we should update this playwright webServer command to use `bos-workspace dev -g ./dist` instead of `yarn start`

**Comment by petersalomonsen** [comment]
> @petersalomonsen yeah, good point on a test for validating hot reload.
> 
> Since the socket server for hot reload needs to be on the same port as the app, it's been difficult to write a test for this -- we've tried mocking web socket or proxying requests, but it was challenging -- I'm not sure, @bb-face may have gotten closer to figuring it out?
> 
> Regardless, I'm wondering if this test belongs in bos-workspace repository instead? Or maybe we should update this playwright webServer command to use `bos-workspace dev -g ./dist` instead of `yarn start`

yeah, this project does not have a dev-server ( and it was never the intention either), but bos-workspace does have that Is we want to build a development environment like bos-workspace in near-bos-webcomponent, then we should also introduce a server for serving the components and a websocket for hot reload.

The main purpose of this project is to be able to serve BOS content inside a custom element in any web page ( not just React apps ). 

It's hard to introduce hot-reload without actually also taking over the role that bos-workspace has as a dev server. So not sure what was the intention behind introducing hot-reload here?

**Comment by petersalomonsen** [comment]
@bb-face @elliotBraem I read this again, and maybe I missed your point. Replacing the `http-server` with ` bos-workspace dev -g ./dist` will probably give you the websocket served from the same port as you suggest. But it would still be good to have the test in this repo, since it feels a bit wrong to have functionality in here that we don't know how to test automatically.

**Comment by bb-face** [comment]
Hey @petersalomonsen! Thanks for the feedback, I just pushed a working test that in my opinion makes sure that hot-reload is works properly, let me know what do you think! The only problem left to fix is the port, unfortuntaely I couldn't find a way to leave the socket server run (during the test) on the same port as the web application but it's a requirement...if you have any suggestions they are more than welcomed!

**Comment by elliotBraem** [comment]
@bb-face, let's move this "enablehotreload" feature into a more generic attribute named "config". This generic attribute can support hot reload as well as other features we may want to configure in the futre, such as "enableComponentSrcDataKey", "defaultFinality", "commitModalBypass", "bypassTransactionConfirmation" which are all used by the VM and we may want to support on this roadmap. For example:

```
 {
    {
      dev: {
        hotreload: {
          enabled: true,
          wss: "wss://localhost:3001",
        },
      },
      features: {
        enableComponentSrcDataKey: true,
        commitModalBypass: {
          authorIds: ["mob.near", "root.near"],
          sources: [
            "cool.near/widget/CoolComponent",
            "awesome.near/widget/AwesomeComponent",
          ],
        },
        bypassTransactionConfirmation: true
        enableWidgetSrcWithCodeOverride: true
      },
      defaultFinality: undefined
    }
  }
```

I am in no way attached to this structure, but notice how hot reload has a customizable source, which we can use with @bb-face's existing test -- otherwise it defaults to the same port. Don't implement the other features, only hot reload support.

**Comment by elliotBraem** [comment]
This has been updated to use the config object with a test that confirms a working hot reload via socket server

**Pull request merged**